core/tx_list.go

```diff
@@ -0,0 +1,332 @@
+	return &txList{
+		strict:  strict,
+		items:   make(map[uint64]*types.Transaction),
+		first:   math.MaxUint64,
+		index:   &nonceHeap{},
+		costcap: new(big.Int),
+	}
+}
+
+// Add tries to inserts a new transaction into the list, returning whether the
+// transaction was acceped, and if yes, any previous transaction it replaced.
+//
+// In case of strict lists (contiguous nonces) the nonce boundaries are updated
+// appropriately with the new transaction. Otherwise (gapped nonces) the heap of
+// nonces is expanded with the new transaction.
+func (l *txList) Add(tx *types.Transaction) (bool, *types.Transaction) {
+	// If an existing transaction is better, discard new one
+	nonce := tx.Nonce()
+
+	old, ok := l.items[nonce]
+	if ok && old.GasPrice().Cmp(tx.GasPrice()) >= 0 {
+		return false, nil
+	}
+	// Otherwise insert the transaction and replace any previous one
+	l.items[nonce] = tx
+	if cost := tx.Cost(); l.costcap.Cmp(cost) < 0 {
+		l.costcap = cost
+	}
+	if l.strict {
+		// In strict mode, maintain the nonce sequence boundaries
+		if nonce < l.first {
+			l.first = nonce
+		}
+		if nonce > l.last {
+			l.last = nonce
+		}
+	} else {
+		// In gapped mode, maintain the nonce heap
+		heap.Push(l.index, nonce)
+	}
+	l.cache = nil
+
+	return true, old
+}
+
+// Forward removes all transactions from the list with a nonce lower than the
+// provided threshold. Every removed transaction is returned for any post-removal
+// maintenance.
+func (l *txList) Forward(threshold uint64) types.Transactions {
+	var removed types.Transactions
+
+	if l.strict {
+		// In strict mode, push the lowest nonce forward to the threshold
+		for l.first < threshold {
+			if tx, ok := l.items[l.first]; ok {
+				removed = append(removed, tx)
+			}
+			delete(l.items, l.first)
+			l.first++
+		}
+		if l.first > l.last {
+			l.last = l.first
+		}
+	} else {
+		// In gapped mode, pop off heap items until the threshold is reached
+		for l.index.Len() > 0 && (*l.index)[0] < threshold {
+			nonce := heap.Pop(l.index).(uint64)
+			removed = append(removed, l.items[nonce])
+			delete(l.items, nonce)
+		}
+	}
+	l.cache = nil
+
+	return removed
+}
+
+// Filter removes all transactions from the list with a cost higher than the
+// provided threshold. Every removed transaction is returned for any post-removal
+// maintenance. Strict-mode invalidated transactions are also returned.
+//
+// This method uses the cached costcap to quickly decide if there's even a point
+// in calculating all the costs or if the balance covers all. If the threshold is
+// loewr than the costcap, the costcap will be reset to a new high after removing
+// expensive the too transactions.
+func (l *txList) Filter(threshold *big.Int) (types.Transactions, types.Transactions) {
+	// If all transactions are blow the threshold, short circuit
+	if l.costcap.Cmp(threshold) <= 0 {
+		return nil, nil
+	}
+	l.costcap = new(big.Int).Set(threshold) // Lower the cap to the threshold
```

Comment:
No. `costcap` is the cached cost of the highest costing transaction. In the lines above this, if the balance of the account is larger than the most expensive transaction, we don't need to filter because everything is already below this.

However if the account balance (threshold) dropped below that of the most expensive cached transaction, then we need to throw out all transactions that suddenly became non executable, so we reduce the cap to the threshold and then filter out everything that's above.